src/Segwit.mo

```diff
@@ -0,0 +1,152 @@
+import Nat32 "mo:base/Nat32";
+import Result "mo:base/Result";
+import Buffer "mo:base/Buffer";
+import Iter "mo:base/Iter";
+import P "mo:base/Prelude";
+import Bech32 "../src/Bech32";
+
+module {
+
+  public type WitnessProgram = {
+    version : Nat8;
+    program : [Nat8];
+  };
+
+  // Convert a Witness Program to a SegWit Address.
+  public func encode(hrp : Text, {version; program} : WitnessProgram) :
+    Result.Result<Text, Text> {
+
+    let bech32Input = Buffer.Buffer<Nat8>(program.size());
+    bech32Input.add(version);
+
+    switch (convertBits(program.vals(), bech32Input, 8, 5, true)) {
+      case (#err (msg)) {
+        return #err (msg);
+      };
+      case _ {};
+    };
+
+    let encoding : Bech32.Encoding = if (version > 0) {
+      #BECH32M;
+    } else {
+      #BECH32;
+    };
+
+    let bech32Result : Text = Bech32.encode(hrp,
+      bech32Input.toArray(), encoding);
+
+    return switch (decode(hrp, bech32Result)) {
+      case (#ok (_)) {
+        #ok (bech32Result)
+      };
+      case (#err (msg)) {
+        #err (msg)
+      };
+    };
+  };
+
+  // Convert a segwit address into a Witness Program.
+  // Verifies the address contains the expected human-readable part and decodes
+  // using Bech32.
+  public func decode(hrp : Text, address : Text)
+    : Result.Result<WitnessProgram, Text> {
+    let (encoding, decodedHrp, data) = switch (Bech32.decode(address)) {
+      case (#ok res) {
+        res
+      };
+      case (#err msg) {
+        return #err (msg);
+      };
+    };
+
+    if (decodedHrp != hrp) {
+      return #err ("Mismatching HRPs");
+    };
+
+    if (data.size() == 0 or data.size() > 65) {
+      return #err ("Invalid data length.");
+    };
+
+    // Split into version and program.
+    let dataIter : Iter.Iter<Nat8> = data.vals();
+    let version : Nat8 = switch (dataIter.next()) {
+      case (?val) {
+        val;
+      };
+      case _ {
+        P.unreachable();
+      };
+    };
+
+    let convertedData = Buffer.Buffer<Nat8>(0);
+    switch (convertBits(dataIter, convertedData, 5, 8, false)) {
+      case (#ok) {
+        let convertedDataSize : Nat = convertedData.size();
+
+        if (convertedDataSize < 2 or convertedDataSize > 40) {
+          return #err ("Wrong output size.")
+        };
+
+        if (data[0] > 16) {
+          return #err ("Invalid witness version.");
+        };
+
+        if (data[0] == 0 and convertedDataSize != 20
+          and convertedDataSize != 32) {
+          return #err ("Program size does not match witness version.");
+        };
+
+        if (data[0] == 0 and encoding != #BECH32 or
+          data[0] != 0 and encoding != #BECH32M) {
+          return #err ("Encoding does not match witness version.");
+        };
+
+        return #ok { version = data[0]; program = convertedData.toArray() };
+      };
+      case _ {
+        return #err("Convert bits failed.");
+      };
+    };
+  };
+
+  // Convert between two bases that are power of 2.
```

Comment:
While it is improbable that the encoding and decoding work if this conversion function is wrong, it would still be nice to have a unit test for this function as well.

Reply:
This means the function becomes public, is that okay?

Reply:
Good point. Is there something like pub(crate) in Motoko?
I would avoid making the function public unnecessarily.

Reply:
Nope, seems our only options are either `public` or `private` (default).